File: PMCRMS/backend/PMCRMS.API/Services/EEStage2WorkflowService.cs
using Microsoft.EntityFrameworkCore;
using PMCRMS.API.Data;
using PMCRMS.API.Models;

namespace PMCRMS.API.Services
{
    /// <summary>
    /// Service for EE Stage 2 workflow (Digital Signature on Certificate)
    /// Handles applications after clerk processing
    /// Status progression: ProcessedByClerk (18) → UnderDigitalSignatureByEE2 (19) → DigitalSignatureCompletedByEE2 (20) → forwards to CE Stage 2 (21)
    /// </summary>
    public class EEStage2WorkflowService
    {
        private readonly PMCRMSDbContext _context;
        private readonly ILogger<EEStage2WorkflowService> _logger;
        private readonly IEmailService _emailService;
        private readonly IHsmService _hsmService;
        private readonly IAutoAssignmentService _autoAssignmentService;
        private readonly IConfiguration _configuration;

        public EEStage2WorkflowService(
            PMCRMSDbContext context,
            ILogger<EEStage2WorkflowService> logger,
            IEmailService emailService,
            IHsmService hsmService,
            IAutoAssignmentService autoAssignmentService,
            IConfiguration configuration)
        {
            _context = context;
            _logger = logger;
            _emailService = emailService;
            _hsmService = hsmService;
            _autoAssignmentService = autoAssignmentService;
            _configuration = configuration;
        }

        /// <summary>
        /// Get all pending applications for EE Stage 2 signature (EXECUTIVE_ENGINEER_SIGN_PENDING status)
        /// </summary>
        public async Task<List<EEStage2ApplicationDto>> GetPendingApplicationsAsync()
        {
            try
            {
                _logger.LogInformation("[EEStage2Workflow] Getting pending applications (EXECUTIVE_ENGINEER_SIGN_PENDING status)");

                var applications = await _context.PositionApplications
                    .Include(a => a.Addresses)
                    .Where(a => a.Status == ApplicationCurrentStatus.EXECUTIVE_ENGINEER_SIGN_PENDING)
                    .OrderBy(a => a.ClerkApprovalDate ?? a.UpdatedDate)
                    .ToListAsync();

                // Get payment amounts from Challan table
                var applicationIds = applications.Select(a => a.Id).ToList();
                var challans = await _context.Challans
                    .Where(c => applicationIds.Contains(c.ApplicationId))
                    .ToDictionaryAsync(c => c.ApplicationId, c => c.Amount);

                var result = applications.Select(a =>
                {
                    // Get permanent address
                    var permanentAddress = a.Addresses.FirstOrDefault(addr => addr.AddressType == "Permanent");
                    var addressText = permanentAddress != null
                        ? $"{permanentAddress.AddressLine1}, {permanentAddress.AddressLine2}, {permanentAddress.AddressLine3}".TrimEnd(',', ' ')
                        : "";

                    return new EEStage2ApplicationDto
                    {
                        Id = a.Id,
                        ApplicationNumber = a.ApplicationNumber ?? "",
                        ApplicantName = $"{a.FirstName} {(string.IsNullOrEmpty(a.MiddleName) ? "" : a.MiddleName + " ")}{a.LastName}".Trim(),
                        ApplicantEmail = a.EmailAddress,
                        PositionType = a.PositionType.ToString(),
                        PropertyAddress = addressText,
                        PaymentAmount = challans.TryGetValue(a.Id, out var amount) ? amount : null,
                        ProcessedByClerkDate = a.ClerkApprovalDate ?? a.UpdatedDate ?? a.CreatedDate,
                        CreatedAt = a.CreatedDate,
                        UpdatedAt = a.UpdatedDate ?? a.CreatedDate
                    };
                }).ToList();

                _logger.LogInformation($"[EEStage2Workflow] Found {result.Count} pending applications");

                return result;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "[EEStage2Workflow] Error getting pending applications");
                throw;
            }
        }

        /// <summary>
        /// Get application details for EE Stage 2 review
        /// </summary>
        public async Task<EEStage2ApplicationDetailDto?> GetApplicationDetailsAsync(int applicationId)
        {
            try
            {
                _logger.LogInformation($"[EEStage2Workflow] Getting application details: {applicationId}");

                var application = await _context.PositionApplications
                    .Include(a => a.Addresses)
                    .Where(a => a.Id == applicationId)
                    .FirstOrDefaultAsync();

                if (application == null) return null;

                // Get challan for payment amount
                var challan = await _context.Challans
                    .Where(c => c.ApplicationId == applicationId)
                    .FirstOrDefaultAsync();

                // Get permanent address
                var permanentAddress = application.Addresses.FirstOrDefault(addr => addr.AddressType == "Permanent");
                var addressText = permanentAddress != null
                    ? $"{permanentAddress.AddressLine1}, {permanentAddress.AddressLine2}, {permanentAddress.AddressLine3}".TrimEnd(',', ' ')
                    : "";

                var detail = new EEStage2ApplicationDetailDto
                {
                    Id = application.Id,
                    ApplicationNumber = application.ApplicationNumber ?? "",
                    ApplicantName = $"{application.FirstName} {(string.IsNullOrEmpty(application.MiddleName) ? "" : application.MiddleName + " ")}{application.LastName}".Trim(),
                    ApplicantEmail = application.EmailAddress,
                    PositionType = application.PositionType.ToString(),
                    PropertyAddress = addressText,
                    CurrentStatus = application.Status.ToString(),
                    PaymentAmount = challan?.Amount,
                    ProcessedByClerkDate = application.ClerkApprovalDate ?? application.UpdatedDate ?? application.CreatedDate,
                    CertificateNumber = null, // Will be generated after signature
                    StatusHistoryCount = 0, // Can be implemented later if needed
                    CreatedAt = application.CreatedDate,
                    UpdatedAt = application.UpdatedDate ?? application.CreatedDate
                };

                return detail;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, $"[EEStage2Workflow] Error getting application details: {applicationId}");
                throw;
            }
        }

        /// <summary>
        /// Generate OTP for digital signature
        /// </summary>
        public async Task<EEStage2OtpResult> GenerateOtpAsync(int applicationId, int eeUserId)
        {
            try
            {
                _logger.LogInformation($"[EEStage2Workflow] Generating OTP for application {applicationId} by EE {eeUserId}");

                var application = await _context.PositionApplications.FindAsync(applicationId);
                if (application == null)
                {
                    return new EEStage2OtpResult
                    {
                        Success = false,
                        Message = "Application not found"
                    };
                }

                if (application.Status != ApplicationCurrentStatus.EXECUTIVE_ENGINEER_SIGN_PENDING)
                {
                    return new EEStage2OtpResult
                    {
                        Success = false,
                        Message = $"Application is not in EXECUTIVE_ENGINEER_SIGN_PENDING status. Current status: {application.Status}"
                    };
                }

                // Fetch the license certificate from database
                var licenseCertificate = await _context.SEDocuments
                    .Where(d => d.ApplicationId == applicationId && d.DocumentType == SEDocumentType.LicenceCertificate)
                    .OrderByDescending(d => d.CreatedDate)
                    .FirstOrDefaultAsync();

                if (licenseCertificate == null || licenseCertificate.FileContent == null)
                {
                    return new EEStage2OtpResult
                    {
                        Success = false,
                        Message = "License certificate not found. Please ensure the certificate has been generated."
                    };
                }

                // Get EE officer and KeyLabel for OTP generation
                var eeOfficer = await _context.Officers.FindAsync(eeUserId);
                if (eeOfficer == null)
                {
                    return new EEStage2OtpResult
                    {
                        Success = false,
                        Message = "Executive Engineer not found"
                    };
                }

                var keyLabel = _configuration[$"HSM:KeyLabels:ExecutiveEngineer"];

                if (string.IsNullOrEmpty(keyLabel))
                {
                    return new EEStage2OtpResult
                    {
                        Success = false,
                        Message = "KeyLabel not configured for ExecutiveEngineer role"
                    };
                }

                _logger.LogInformation(
                    "Generating OTP for EE Stage 2 using KeyLabel '{KeyLabel}' for officer {OfficerId}",
                    keyLabel, eeUserId);

                // Generate OTP using HSM
                var otpResult = await _hsmService.GenerateOtpAsync(
                    transactionId: applicationId.ToString(),
                    keyLabel: keyLabel,
                    otpType: "single"
                );
                
                if (!otpResult.Success)
                {
                    return new EEStage2OtpResult
                    {
                        Success = false,
                        Message = $"Failed to generate OTP: {otpResult.ErrorMessage}"
                    };
                }

                _logger.LogInformation("✅ [EEStage2Workflow] OTP generated and sent by HSM for application {ApplicationId}: {Message}",
                    applicationId, otpResult.Message);

                return new EEStage2OtpResult
                {
                    Success = true,
                    Message = otpResult.Message ?? "OTP sent successfully. Please check your registered mobile/email for the OTP code.",
                    OtpReference = applicationId.ToString()
                };
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, $"[EEStage2Workflow] Error generating OTP for application {applicationId}");
                return new EEStage2OtpResult
                {
                    Success = false,
                    Message = $"Error generating OTP: {ex.Message}"
                };
            }
        }

        /// <summary>
        /// Apply digital signature to certificate
        /// Updates status from EXECUTIVE_ENGINEER_SIGN_PENDING (32) to CITY_ENGINEER_SIGN_PENDING (34)
        /// </summary>
        public async Task<EEStage2SignResult> ApplyDigitalSignatureAsync(int applicationId, int eeUserId, string otpCode)
        {
            try
            {
                _logger.LogInformation($"[EEStage2Workflow] Applying digital signature for application {applicationId} by EE {eeUserId}");

                var application = await _context.PositionApplications
                    .FirstOrDefaultAsync(a => a.Id == applicationId);

                if (application == null)
                {
                    return new EEStage2SignResult
                    {
                        Success = false,
                        Message = "Application not found"
                    };
                }

                if (application.Status != ApplicationCurrentStatus.EXECUTIVE_ENGINEER_SIGN_PENDING)
                {
                    return new EEStage2SignResult
                    {
                        Success = false,
                        Message = $"Application is not in EXECUTIVE_ENGINEER_SIGN_PENDING status. Current status: {application.Status}"
                    };
                }

                // Get the officer's username for audit
                var eeOfficer = await _context.Officers.FindAsync(eeUserId);
                if (eeOfficer == null)
                {
                    return new EEStage2SignResult
                    {
                        Success = false,
                        Message = "Executive Engineer not found."
                    };
                }

                // First, get the license certificate document from database
                var licenseCertificate = await _context.SEDocuments
                    .FirstOrDefaultAsync(d => d.ApplicationId == applicationId && 
                                            d.DocumentType == SEDocumentType.LicenceCertificate);

                if (licenseCertificate == null || licenseCertificate.FileContent == null)
                {
                    return new EEStage2SignResult
                    {
                        Success = false,
                        Message = "License certificate not found. Certificate must be generated before signing."
                    };
                }

                // Get KeyLabel for EE from configuration
                var keyLabel = _configuration[$"HSM:KeyLabels:ExecutiveEngineer"];

                if (string.IsNullOrEmpty(keyLabel))
                {
                    return new EEStage2SignResult
                    {
                        Success = false,
                        Message = "KeyLabel not configured for ExecutiveEngineer role"
                    };
                }

                _logger.LogInformation(
                    "Signing license certificate for EE Stage 2 using KeyLabel '{KeyLabel}' for officer {OfficerId}",
                    keyLabel, eeUserId);

                // Get EE signature coordinates from configuration for license certificate
                var eeCoordinates = _configuration["HSM:LicenseCertificateSignatureCoordinates:ExecutiveEngineer"] 
                    ?? "100,50,250,110,1";
                
                _logger.LogInformation(
                    "📍 Signature coordinates for {Role} on license certificate: {Coordinates}",
                    eeOfficer.Role, eeCoordinates);

                // Convert PDF to Base64
                var base64Pdf = Convert.ToBase64String(licenseCertificate.FileContent);

                // Sign PDF using HSM
                var signRequest = new HsmSignRequest
                {
                    TransactionId = applicationId.ToString(),
                    KeyLabel = keyLabel,
                    Base64Pdf = base64Pdf,
                    Otp = otpCode,
                    Coordinates = eeCoordinates,
                    PageLocation = "last",
                    OtpType = "single"
                };

                var signResult = await _hsmService.SignPdfAsync(signRequest);

                if (!signResult.Success)
                {
                    _logger.LogError("HSM signature failed for application {ApplicationId}: {Error}", 
                        applicationId, signResult.ErrorMessage);
                    
                    return new EEStage2SignResult
                    {
                        Success = false,
                        Message = $"Digital signature failed: {signResult.ErrorMessage}"
                    };
                }

                if (string.IsNullOrEmpty(signResult.SignedPdfBase64))
                {
                    return new EEStage2SignResult
                    {
                        Success = false,
                        Message = "Signed PDF not returned by HSM"
                    };
                }

                // Convert Base64 back to bytes and update database
                var signedPdfBytes = Convert.FromBase64String(signResult.SignedPdfBase64);
                licenseCertificate.FileContent = signedPdfBytes;
                licenseCertificate.FileSize = (decimal)(signedPdfBytes.Length / 1024.0);
                licenseCertificate.UpdatedDate = DateTime.UtcNow;
                
                _logger.LogInformation(
                    "[EEStage2Workflow] Updated license certificate FileContent in database ({Size} KB) for application {ApplicationId}",
                    licenseCertificate.FileSize, applicationId);

                // Update application status to CITY_ENGINEER_SIGN_PENDING and mark EE Stage 2 signature complete
                application.Status = ApplicationCurrentStatus.CITY_ENGINEER_SIGN_PENDING;
                application.EEStage2DigitalSignatureApplied = true;
                application.EEStage2DigitalSignatureDate = DateTime.UtcNow;
                application.EEStage2ApprovalStatus = true;
                application.EEStage2ApprovalDate = DateTime.UtcNow;
                application.UpdatedDate = DateTime.UtcNow;

                _logger.LogInformation($"[EEStage2Workflow] BEFORE SaveChanges - Application {applicationId} Status: {application.Status} (Enum Value: {(int)application.Status}), Certificate bytes: {licenseCertificate.FileContent?.Length ?? 0}");

                await _context.SaveChangesAsync();

                _logger.LogInformation($"[EEStage2Workflow] AFTER SaveChanges - Application {applicationId} Status: {application.Status} (Enum Value: {(int)application.Status}), Digital signature applied successfully");
                
                // Verify the status was actually saved by re-reading from database
                var savedApplication = await _context.PositionApplications.AsNoTracking().FirstOrDefaultAsync(a => a.Id == applicationId);
                _logger.LogInformation($"[EEStage2Workflow] VERIFICATION - Re-read from DB - Application {applicationId} Status: {savedApplication?.Status} (Enum Value: {(int)(savedApplication?.Status ?? 0)}), AssignedCEStage2Id: {savedApplication?.AssignedCEStage2Id}");


                // Auto-assign to City Engineer for final signature
                try
                {
                    _logger.LogInformation($"[EEStage2Workflow] Attempting auto-assignment to City Engineer for application {applicationId}");
                    
                    var assignmentResult = await _autoAssignmentService.AutoAssignToNextWorkflowStageAsync(
                        applicationId,
                        ApplicationCurrentStatus.CITY_ENGINEER_SIGN_PENDING,
                        eeUserId
                    );

                    if (assignmentResult != null)
                    {
                        _logger.LogInformation(
                            "[EEStage2Workflow] ✅ Application {ApplicationId} auto-assigned to City Engineer (Officer ID: {OfficerId}) for final signature",
                            applicationId,
                            assignmentResult.AssignedToOfficerId
                        );
                        
                        // Re-verify the assignment was saved
                        var verifyAssignment = await _context.PositionApplications.AsNoTracking()
                            .Where(a => a.Id == applicationId)
                            .Select(a => new { a.Id, a.Status, a.AssignedCEStage2Id, a.AssignedToCEStage2Date })
                            .FirstOrDefaultAsync();
                        _logger.LogInformation(
                            "[EEStage2Workflow] Assignment verification - App {AppId}: Status={Status}, AssignedCEStage2Id={CEId}, AssignedDate={Date}",
                            verifyAssignment?.Id, verifyAssignment?.Status, verifyAssignment?.AssignedCEStage2Id, verifyAssignment?.AssignedToCEStage2Date
                        );
                    }
                    else
                    {
                        _logger.LogWarning(
                            "[EEStage2Workflow] ⚠️ Application {ApplicationId} status updated to CITY_ENGINEER_SIGN_PENDING but no City Engineer available for auto-assignment. Application will appear as unassigned on CE dashboard.",
                            applicationId
                        );
                    }
                }
                catch (Exception assignEx)
                {
                    _logger.LogError(assignEx, "[EEStage2Workflow] ❌ Failed to auto-assign to City Engineer for application {ApplicationId}. Application will appear as unassigned on CE dashboard.", applicationId);
                    // Don't fail the signature operation if assignment fails
                }

                // Send email notification (non-blocking)
                try
                {
                    var viewUrl = $"{GetBaseUrl()}/view-application/{applicationId}";
                    // TODO: Implement SendEE2SignatureCompletedEmailAsync in EmailService
                    // await _emailService.SendEE2SignatureCompletedEmailAsync(
                    //     application.EmailAddress,
                    //     $"{application.FirstName} {application.LastName}",
                    //     application.ApplicationNumber,
                    //     viewUrl
                    // );
                }
                catch (Exception emailEx)
                {
                    _logger.LogError(emailEx, $"[EEStage2Workflow] Failed to send signature completion email for application {applicationId}");
                }

                return new EEStage2SignResult
                {
                    Success = true,
                    Message = "Digital signature applied successfully. Application forwarded to City Engineer for final approval.",
                    ApplicationId = applicationId,
                    NewStatus = ApplicationCurrentStatus.CITY_ENGINEER_SIGN_PENDING.ToString(),
                    SignedCertificateUrl = $"/api/Download/certificate/{applicationId}"
                };
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, $"[EEStage2Workflow] Error applying digital signature for application {applicationId}");
                return new EEStage2SignResult
                {
                    Success = false,
                    Message = $"Error applying digital signature: {ex.Message}"
                };
            }
        }

        /// <summary>
        /// Get completed applications (signed by EE Stage 2)
        /// </summary>
        public async Task<List<EEStage2ApplicationDto>> GetCompletedApplicationsAsync()
        {
            try
            {
                _logger.LogInformation("[EEStage2Workflow] Getting completed applications");

                var applications = await _context.PositionApplications
                    .Include(a => a.Addresses)
                    .Where(a => a.EEStage2DigitalSignatureApplied == true)
                    .OrderByDescending(a => a.EEStage2DigitalSignatureDate)
                    .ToListAsync();

                // Get payment amounts from Challan table
                var applicationIds = applications.Select(a => a.Id).ToList();
                var challans = await _context.Challans
                    .Where(c => applicationIds.Contains(c.ApplicationId))
                    .ToDictionaryAsync(c => c.ApplicationId, c => c.Amount);

                var result = applications.Select(a =>
                {
                    // Get permanent address
                    var permanentAddress = a.Addresses.FirstOrDefault(addr => addr.AddressType == "Permanent");
                    var addressText = permanentAddress != null
                        ? $"{permanentAddress.AddressLine1}, {permanentAddress.AddressLine2}, {permanentAddress.AddressLine3}".TrimEnd(',', ' ')
                        : "";

                    return new EEStage2ApplicationDto
                    {
                        Id = a.Id,
                        ApplicationNumber = a.ApplicationNumber ?? "",
                        ApplicantName = $"{a.FirstName} {(string.IsNullOrEmpty(a.MiddleName) ? "" : a.MiddleName + " ")}{a.LastName}".Trim(),
                        ApplicantEmail = a.EmailAddress,
                        PositionType = a.PositionType.ToString(),
                        PropertyAddress = addressText,
                        PaymentAmount = challans.TryGetValue(a.Id, out var amount) ? amount : null,
                        ProcessedByClerkDate = a.ClerkApprovalDate ?? a.UpdatedDate ?? a.CreatedDate,
                        CreatedAt = a.CreatedDate,
                        UpdatedAt = a.UpdatedDate ?? a.CreatedDate
                    };
                }).ToList();

                _logger.LogInformation($"[EEStage2Workflow] Found {result.Count} completed applications");

                return result;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "[EEStage2Workflow] Error getting completed applications");
                throw;
            }
        }

        private string GetBaseUrl()
        {
            return Environment.GetEnvironmentVariable("FRONTEND_URL") ?? throw new InvalidOperationException("FRONTEND_URL environment variable not configured");
        }
    }

    // DTOs for EE Stage 2 Workflow
    public class EEStage2ApplicationDto
    {
        public int Id { get; set; }
        public string ApplicationNumber { get; set; } = string.Empty;
        public string ApplicantName { get; set; } = string.Empty;
        public string ApplicantEmail { get; set; } = string.Empty;
        public string PositionType { get; set; } = string.Empty; // Changed from ApplicationType to match frontend
        public string PropertyAddress { get; set; } = string.Empty;
        public decimal? PaymentAmount { get; set; }
        public DateTime ProcessedByClerkDate { get; set; }
        public DateTime CreatedAt { get; set; }
        public DateTime UpdatedAt { get; set; }
    }

    public class EEStage2ApplicationDetailDto : EEStage2ApplicationDto
    {
        public string CurrentStatus { get; set; } = string.Empty;
        public string? CertificateNumber { get; set; }
        public int StatusHistoryCount { get; set; }
    }

    public class EEStage2OtpResult
    {
        public bool Success { get; set; }
        public string Message { get; set; } = string.Empty;
        public string? OtpReference { get; set; }
    }

    public class EEStage2SignResult
    {
        public bool Success { get; set; }
        public string Message { get; set; } = string.Empty;
        public int ApplicationId { get; set; }
        public string NewStatus { get; set; } = string.Empty;
        public string? SignedCertificateUrl { get; set; }
    }

    public class EEStage2SignRequest
    {
        public string OtpCode { get; set; } = string.Empty;
    }
}

File: PMCRMS/frontend/src/services/eeStage2WorkflowService.ts
/**
 * EE Stage 2 Workflow Service
 * Handles API calls for Executive Engineer Stage 2 Certificate Digital Signature workflow
 */

import axios from "axios";
import { getApiUrl, getToken } from "./apiClient";

// DTOs matching backend
export interface EEStage2ApplicationDto {
  id: number; // Backend uses 'id'
  applicationId?: number; // For compatibility
  applicationNumber: string;
  applicantName: string;
  applicantEmail: string;
  positionType: string; // Backend property name
  buildingType?: string; // Legacy/alternative property
  propertyAddress: string;
  paymentAmount?: number;
  processedByClerkDate?: string; // Backend property name
  clerkProcessedDate?: string; // Legacy/alternative property
  plotArea?: number; // Legacy property
  district?: string; // Legacy property
  clerkRemarks?: string;
  paymentDate?: string;
  paymentReference?: string;
  createdAt: string;
  updatedAt: string;
}

export interface EEStage2ApplicationDetailDto {
  applicationId: number;
  applicationNumber: string;
  applicantName: string;
  applicantEmail: string;
  applicantContact: string;
  buildingType: string;
  plotArea: number;
  buildingArea: number;
  floors: number;
  district: string;
  taluka: string;
  village: string;
  clerkProcessedDate?: string;
  clerkRemarks?: string;
  paymentAmount: number;
  paymentDate?: string;
  paymentReference: string;
  certificateGeneratedDate?: string;
  certificatePath?: string;
  currentStatus: string;
  jeApprovedDate?: string;
  jeOfficerName?: string;
  aeApprovedDate?: string;
  aeOfficerName?: string;
  ee1ApprovedDate?: string;
  ee1OfficerName?: string;
  ce1ApprovedDate?: string;
  ce1OfficerName?: string;
}

export interface EEStage2OtpResult {
  success: boolean;
  message: string;
  otpReference?: string;
}

export interface EEStage2SignResult {
  success: boolean;
  message: string;
  applicationId?: number;
  newStatus?: string;
  signedCertificateUrl?: string;
}

export interface EEStage2SignRequest {
  otpCode: string;
  comments?: string;
}

export interface EEStage2Statistics {
  pendingCount: number;
  completedCount: number;
  todayProcessed: number;
  weekProcessed: number;
  monthProcessed: number;
}

class EEStage2WorkflowService {
  /**
   * Get pending applications for EE Stage 2 (status 18 - ProcessedByClerk)
   */
  async getPendingApplications(): Promise<EEStage2ApplicationDto[]> {
    try {
      const token = getToken();
      const response = await axios.get(`${getApiUrl()}/EEStage2/Pending`, {
        headers: {
          Authorization: `Bearer ${token}`,
        },
      });
      if (response.data.success) {
        return response.data.data as EEStage2ApplicationDto[];
      }
      throw new Error(
        response.data.message || "Failed to fetch pending applications"
      );
    } catch (error: unknown) {
      if (axios.isAxiosError(error)) {
        throw new Error(
          error.response?.data?.message || error.message || "Network error"
        );
      }
      throw error;
    }
  }

  /**
   * Get completed applications by EE Stage 2
   */
  async getCompletedApplications(): Promise<EEStage2ApplicationDto[]> {
    try {
      const token = getToken();
      const response = await axios.get(`${getApiUrl()}/EEStage2/Completed`, {
        headers: {
          Authorization: `Bearer ${token}`,
        },
      });
      if (response.data.success) {
        return response.data.data as EEStage2ApplicationDto[];
      }
      throw new Error(
        response.data.message || "Failed to fetch completed applications"
      );
    } catch (error: unknown) {
      if (axios.isAxiosError(error)) {
        throw new Error(
          error.response?.data?.message || error.message || "Network error"
        );
      }
      throw error;
    }
  }

  /**
   * Get detailed application information
   */
  async getApplicationDetails(
    applicationId: number
  ): Promise<EEStage2ApplicationDetailDto> {
    try {
      const token = getToken();
      const response = await axios.get(
        `${getApiUrl()}/EEStage2/Application/${applicationId}`,
        {
          headers: {
            Authorization: `Bearer ${token}`,
          },
        }
      );
      if (response.data.success) {
        return response.data.data as EEStage2ApplicationDetailDto;
      }
      throw new Error(
        response.data.message || "Failed to fetch application details"
      );
    } catch (error: unknown) {
      if (axios.isAxiosError(error)) {
        throw new Error(
          error.response?.data?.message || error.message || "Network error"
        );
      }
      throw error;
    }
  }

  /**
   * Generate OTP for digital signature via HSM
   */
  async generateOtp(applicationId: number): Promise<EEStage2OtpResult> {
    try {
      const token = getToken();
      const response = await axios.post(
        `${getApiUrl()}/EEStage2/GenerateOtp/${applicationId}`,
        {},
        {
          headers: {
            Authorization: `Bearer ${token}`,
          },
        }
      );
      if (response.data.success) {
        return {
          success: true,
          message: response.data.message,
          otpReference: response.data.data?.otpReference,
        };
      }
      throw new Error(response.data.message || "Failed to generate OTP");
    } catch (error: unknown) {
      if (axios.isAxiosError(error)) {
        throw new Error(
          error.response?.data?.message || error.message || "Network error"
        );
      }
      throw error;
    }
  }

  /**
   * Apply digital signature with OTP verification
   */
  async applyDigitalSignature(
    applicationId: number,
    otpCode: string,
    comments?: string
  ): Promise<EEStage2SignResult> {
    try {
      const token = getToken();
      const requestData: EEStage2SignRequest = { otpCode, comments };
      const response = await axios.post(
        `${getApiUrl()}/EEStage2/Sign/${applicationId}`,
        requestData,
        {
          headers: {
            Authorization: `Bearer ${token}`,
            "Content-Type": "application/json",
          },
        }
      );
      if (response.data.success) {
        return {
          success: true,
          message: response.data.message,
          applicationId: response.data.data?.applicationId,
          newStatus: response.data.data?.newStatus,
          signedCertificateUrl: response.data.data?.signedCertificateUrl,
        };
      }
      throw new Error(
        response.data.message || "Failed to apply digital signature"
      );
    } catch (error: unknown) {
      if (axios.isAxiosError(error)) {
        throw new Error(
          error.response?.data?.message || error.message || "Network error"
        );
      }
      throw error;
    }
  }

  /**
   * Get statistics for EE Stage 2 dashboard
   */
  async getStatistics(): Promise<EEStage2Statistics> {
    try {
      const token = getToken();
      const response = await axios.get(`${getApiUrl()}/EEStage2/Statistics`, {
        headers: {
          Authorization: `Bearer ${token}`,
        },
      });
      if (response.data.success) {
        return response.data.data as EEStage2Statistics;
      }
      throw new Error(response.data.message || "Failed to fetch statistics");
    } catch (error: unknown) {
      if (axios.isAxiosError(error)) {
        throw new Error(
          error.response?.data?.message || error.message || "Network error"
        );
      }
      throw error;
    }
  }
}

export default new EEStage2WorkflowService();
